### ISSUE TITLE
Train encoder weights on not labeled data

### ISSUE THREAD

**Issue by VladislavAD** [qubvel/segmentation_models#315]
Hi, I want to make pretrained weights for my task but can't find any information how to do that. My problems are long training time and small amount of labeled images, but I have a lot of unlabeled images and I think that I can make pretrained weights using them for faster learning on labeled data. Is there any way to train only encoder? Maybe I can extract encoder weights from trained model?

**Comment by JordanMakesMaps** [comment]
You can use Keras to train the same encoder that you're using as an encoder for the segmentation models, and create a little function that will take the weights per layer from the keras model, and place them in the correct layer of the segmentation model.

I did the same thing not too long ago and it worked fine.

**Comment by VladislavAD** [comment]
The problem is I can't find approach to actualy train this weights. Can I just train my model using not labeled data and zeros as ground-truth and get trained encoder from it, would it be correct? As I understand, encoder perform as feature extractor, but will it be trained properly if all ground-truth values are zeros?

**Comment by JordanMakesMaps** [comment]
Can you provide an example of the data you're working with? How many classes, and how many images do you have that are labeled and unlabeled?

These segmentation architectures are trained using a supervised method, in that, they need samples with labels to learn the difference between classes. Their training data should be in the form of an image, and a segmentation map that describes the class for each pixel in the image.

But the encoder is just a CNN without the fully-connected layers, and qubvel did a good job of making sure the encoders in this repo are the same as the ones in Keras. So what you could do is, if all the images you're working have a single label for an image classification task (e.g. a picture of a dog has a corresponding label of  dog, a picture of a cat has a corresponding label of a cat) then you could train an encoder (ResNet50 or something) on those images as an image classifier, _then_ take the weights from that train model and and place them in the encoder part of the segmentation model.

Note that this _might_ help your segmentation model do better (really, just converge faster), but you still need to have labeled segmentation maps for each image to train a segmentation model. There's no getting around that.

What is keeping you from obtaining segmentation maps? There are free software tools that help you annotate your data.

**Comment by VladislavAD** [comment]
I have very industry-specific task of product defects marking, so I have enormous amout of raw images of product but I need manualy mark each defect because traditional approaches don't give good results. 

I can find defects using U-Net+MobileNetV2 and this approach give decent result but I did't reach prefered accuracy and now I'm slowed down by exponentialy increasing learning time with exponentially decreasing accuracy improvement and look for different approaches how I can make advantage from what I have now. I'm trying to find more representative selection of data to reduce the number of images for faster learning with the same accuracy, I'm trying to find a way to continue learning with additional data (no success so far, train from scratch works better), and I observed that first 300 epochs U-Net looks like training encoder, because this epochs I have almost zero iou score over validation data, and then comes jump to ~0.4 iou score, so I think that pre-trained encoder will allow me to skip this 300 epochs. 

As for numbers, I have ~15k selected "squares"(I cut original image into squares because its too big) from labeled images (50% with defects, 50% without defects but this is not all non-defect squares I have) and the number of classes are between 1 and 20, currently working with 4. Also for some reason I get only 30% learning speedup when use Tesla V100 compared to GTX1080, is this strange and I'm doing something wrong?

> So what you could do is, if all the images you're working have a single label for an image classification task (e.g. a picture of a dog has a corresponding label of dog, a picture of a cat has a corresponding label of a cat) then you could train an encoder (ResNet50 or something) on those images as an image classifier, then take the weights from that train model and and place them in the encoder part of the segmentation model.

So I can pick out images with single defect on it, and try to train image classifier instead of image segmentation and then use encoder weights from it as segmentation backbone?
